go/cmd/dolt/commands/ls.go

```diff
@@ -83,198 +80,142 @@ func (cmd LsCmd) Exec(ctx context.Context, commandStr string, args []string, dEn
-			if verr != nil {
-				return verr
-			}
-			cli.Println(ls)
+	var userTables []string
+	var systemTables []string
+	for _, row := range rows {
+		tableName := row[0].(string)
+		if doltdb.HasDoltPrefix(tableName) {
+			systemTables = append(systemTables, tableName)
 		} else {
-			cli.Println("\t", tbl)
+			userTables = append(userTables, tableName)
 		}
 	}
 
-	return nil
-}
-
-func listTableVerbose(ctx context.Context, tbl string, root *doltdb.RootValue) (string, errhand.VerboseError) {
-	h, _, err := root.GetTableHash(ctx, tbl)
-	if err != nil {
-		return "", errhand.BuildDError("error: failed to get table hash").AddCause(err).Build()
+	if apr.Contains(cli.AllFlag) {
+		err = printUserTables(userTables, apr, queryist, sqlCtx)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		}
+		err = printSystemTables(systemTables)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		}
+	} else if apr.Contains(cli.SystemFlag) {
+		err = printSystemTables(systemTables)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		}
+	} else {
+		err = printUserTables(userTables, apr, queryist, sqlCtx)
+		if err != nil {
+			return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		}
 	}
 
-	tblVal, _, err := root.GetTable(ctx, tbl)
-	if err != nil {
-		return "", errhand.BuildDError("error: failed to get table").AddCause(err).Build()
-	}
+	return HandleVErrAndExitCode(nil, usage)
+}
 
-	rows, err := tblVal.GetRowData(ctx)
-	if err != nil {
-		return "", errhand.BuildDError("error: failed to get row data").AddCause(err).Build()
-	}
-	cnt, err := rows.Count()
-	if err != nil {
-		return "", errhand.VerboseErrorFromError(err)
+// constructInterpolatedDoltLsQuery generates the sql query necessary to list all tables and interpolates this query
+// to prevent sql injection
+func constructInterpolatedDoltLsQuery(apr *argparser.ArgParseResults) (string, error) {
+	query := "show tables"
+	if apr.NArg() == 1 {
+		query += " as of ?"
+		interpolatedQuery, err := dbr.InterpolateForDialect(query, []interface{}{apr.Arg(0)}, dialect.MySQL)
+		if err != nil {
+			return "", err
+		}
+		return interpolatedQuery, nil
+	} else {
+		return query, nil
 	}
-
-	return fmt.Sprintf("\t%-32s %s    %d rows\n", tbl, h.String(), cnt), nil
 }
 
-func printSystemTables(ctx context.Context, root *doltdb.RootValue, ddb *doltdb.DoltDB, verbose bool) errhand.VerboseError {
-	perSysTbls, err := doltdb.GetPersistedSystemTables(ctx, root)
-	if err != nil {
-		return errhand.BuildDError("error retrieving persisted table names").AddCause(err).Build()
+func printUserTables(tableNames []string, apr *argparser.ArgParseResults, queryist cli.Queryist, sqlCtx *sql.Context) error {
+	var label string
+	if apr.NArg() == 0 {
+		label = "working set"
+	} else {
+		row, err := GetRowsForSql(queryist, sqlCtx, "select hashof('"+apr.Arg(0)+"')")
```

Comment:
I tend to avoid string addition. I think this is more readable:

		query := fmt.Sprintf("select hashof('%s')", apr.Arg(0))
		row, err := GetRowsForSql(queryist, sqlCtx, query